chia/consensus/multiprocess_validation.py

```diff
@@ -227,9 +228,30 @@ async def pre_validate_blocks_multiprocessing(
     diff_ssis: List[Tuple[uint64, uint64]] = []
     for block in blocks:
         if block.height != 0:
-            assert block_records.contains_block(block.prev_header_hash)
             if prev_b is None:
-                prev_b = block_records.block_record(block.prev_header_hash)
+                prev_b = await block_records.get_block_record_from_db(block.prev_header_hash)
+            assert prev_b is not None
+
+            # the call to block_to_block_record() requires the previous
+            # block is in the cache
+            # and make_sub_epoch_summary() requires all blocks until we find one
+            # that includes a sub_epoch_summary
+            curr = prev_b
+            block_records.add_block_record(curr)
+            counter = 0
+            # TODO: It would probably be better to make
+            # get_next_sub_slot_iters_and_difficulty() async and able to pull
```

Comment:
why do we need this, we are supposed to have al these blocks in the cache, this feels like we are patching some bug in the cache

Reply:
do you know where we are supposed to populate the cache to cover this?

Reply:
what I mean by this is; we clearly have a bug because our current code expects that we never reorg farther back than the cache covers. That's a big part of what this PR addresses. We don't warm up the cache in case of reorgs, the way we normally do when we start a long-sync. I think the long term fix to this would be to make fewer things rely on the cache being pre-populated. Perhaps that's impractical if python always yield when you call an `async` function, regardless if it can return without blocking (is that the case?).